src/test/socket/socket/test_socket.rs

```diff
@@ -0,0 +1,501 @@
+            flag: Cond::Any,
+            protocol: Cond::Only(&[libc::IPPROTO_TCP]),
+            expected_errno: Some(libc::EPROTONOSUPPORT),
+        },
+        // if we use the UDP protocol without the SOCK_DGRAM type
+        ErrorCondition {
+            domain: Cond::Any,
+            sock_type: Cond::Not(&[libc::SOCK_DGRAM]),
+            flag: Cond::Any,
+            protocol: Cond::Only(&[libc::IPPROTO_UDP]),
+            expected_errno: Some(libc::EPROTONOSUPPORT),
+        },
+        // if we use the SOCK_PACKET type
+        ErrorCondition {
+            domain: Cond::Any,
+            sock_type: Cond::Only(&[libc::SOCK_PACKET]),
+            flag: Cond::Any,
+            protocol: Cond::Any,
+            expected_errno: Some(libc::ESOCKTNOSUPPORT),
+        },
+    ];
+
+    let tests = if run_only_passing_tests {
+        get_passing_tests()
+    } else {
+        get_all_tests()
+    };
+
+    if let Err(_) = run_tests(tests.iter(), &error_conditions, summarize) {
+        println!("Failed.");
+        std::process::exit(1);
+    }
+
+    println!("Success.");
+}
+
+fn get_passing_tests() -> Vec<(SocketFn, SocketArguments)> {
+    // the different arguments to try (including invalid args)
+    let domains = [libc::AF_INET, 0xABBA];
+    let sock_types = [libc::SOCK_STREAM, libc::SOCK_DGRAM];
+    let flags = [0, libc::SOCK_NONBLOCK, libc::SOCK_CLOEXEC];
+    let protocols = [0, libc::IPPROTO_TCP, libc::IPPROTO_UDP];
+
+    // product of the sets of arguments
+    let mut tests = Vec::new();
+
+    for socket_fn in [SocketFn::Socket, SocketFn::Syscall].iter() {
+        for domain in domains.iter() {
+            for sock_type in sock_types.iter() {
+                for flag in flags.iter() {
+                    for protocol in protocols.iter() {
+                        let args = SocketArguments {
+                            domain: *domain,
+                            sock_type: *sock_type,
+                            flag: *flag,
+                            protocol: *protocol,
+                        };
+                        tests.push((*socket_fn, args));
+                    }
+                }
+            }
+        }
+    }
+
+    tests
+}
+
+fn get_all_tests() -> Vec<(SocketFn, SocketArguments)> {
+    // the different arguments to try (including invalid args)
+    let domains = [
+        libc::AF_UNIX,
+        libc::AF_LOCAL,
+        libc::AF_INET,
+        //libc::AF_AX25,
+        //libc::AF_IPX,
+        //libc::AF_APPLETALK,
+        //libc::AF_X25,
+        libc::AF_INET6,
+        //libc::AF_DECnet,
+        //libc::AF_KEY,
+        //libc::AF_NETLINK,
+        //libc::AF_PACKET,
+        //libc::AF_RDS,
+        //libc::AF_PPPOX,
+        //libc::AF_LLC,
+        //libc::AF_IB,
+        //libc::AF_MPLS,
+        //libc::AF_CAN,
+        //libc::AF_TIPC,
+        //libc::AF_BLUETOOTH,
+        //libc::AF_ALG,
+        //libc::AF_VSOCK,
+        //libc::AF_XDP,
```

Comment:
It's unclear why some of these are commented out. For example, are we supposed to uncomment them later if we add support? I suggest adding a comment to explain why.

Reply:
This was just to easily see what we aren't testing (I don't think Shadow will ever support `AF_IPX` for example), but are still technically supported by the socket system call. I think I'll remove these comments though as this really isn't that useful.